### ISSUE TITLE
Validation too brittle

### ISSUE THREAD

**Issue by lastcoolnameleft** [Azure/simdem#54]
Validation fails for me having the Azure CLI, because `az --version` lists the modules installed.  As more modules are installed, this might drift and start failing.

In addition, the Python listing could drift: 
```
Python (Darwin) 2.7.10 (default, Feb  7 2017, 00:08:15)
[GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.34)]

Python location '/Users/thfalgou/lib/azure-cli/bin/python'
```

vs 

```
Python (Linux) 3.6.1 (default, Aug 17 2017, 10:48:15)
[GCC 4.8.4]

Python location '/opt/az/bin/python3'
```

I propose adding a validation type that simply checks for error code.  i.e. $?

**Comment by SorraTheOrc** [comment]
Yes, this is a problem in a few places. There is a tension here, I don't want SimDem specific syntax to creep into the markdown because this makes it less readable. However, there are times when the output of a command is simply too noisy for the current crude method.

I'm thinking of two partial solutions to this:

  1. one is to allow wildcards in the results (e.g. `starting text required * ending text required`).
  2. Check for error codes, since any failing command is presumably a failed valiation step.

I think item 2. is what you are proposing. If there is a difference between our thinking it is that I propose we do this on every command and not require the user to enter a Results section at all.

Do you think this will work?

**Comment by SorraTheOrc** [comment]
Workaround:

In the meantime it is possible to change the expected similarity for tests, see "expected_similarity" in https://github.com/Azure/simdem/tree/master/demo_scripts/simdem/test#running-in-test-mode

**Comment by lastcoolnameleft** [comment]
What about using test annotations?

For example:
# Validaton
[//]: <> (test_error_code)

'''
az --version
'''

https://stackoverflow.com/questions/4823468/comments-in-markdown

**Comment by SorraTheOrc** [comment]
Is there any command where a failure exit code is acceptable? My feeling is that we should just test for it by default.

As for your test_annotations suggestion, I think there is something in that.

I like it because it doesn't require the demo writer to know anything about SimDem.

I also like it because it can easily be expanded to accommodate things like regular expressions.

I also like it because we could move the current expected_similarity value into the annotations (it's been bothering me that the code block annotations are "traditionally" used for identifying render information, such as the language used).

However, we need to think a little about the syntax and behavior. If we have a test annotation and a Results section do they both count as tests?

I'd suggestion it would be better to have the test annotation after the command, but that's a fairly minor point. Is there any reason why you put it before?

**Comment by lastcoolnameleft** [comment]
Re: failure exit code.  Good point.  We should test for it by default; however, I want to ensure that we have the ability to test for the exit and only the exit code (i.e. no percentage matching)

The regex Use Case would be great. 

Re: Syntax/behavior.  Perhaps certain commands could require a results section (e.g. percent_match, last_line_match, first_line_match)

Re: before/after the command.  Annotations are usually before the signature, so you're right, let's put it above.

**Comment by SorraTheOrc** [comment]
I propose we close this, it is addressed in the refactoring of SimDem, which has a much more flexible validation model.

**Comment by lastcoolnameleft** [comment]
I feel that this issue is still open as I don't have a clear understanding of how we want to implement flexibility of testing (i.e. syntax, features, etc.)  We've thrown out a few ideas, but I don't believe we've solidified on the details.

Outstanding questions:
* Should we always test for exit code?
  * If so, what should be action for a passed and failed validation?  For example:
    * If passed, should we print anything, or just move on
    * If failed, should we exit?  Should we print an error and move on?
* Do we want to allow wildcards/regex?  That could get tricky, but powerful
* How do we want to customize the expected_similarity?  (i.e. in the language type?  In some other format?)

Anything else I might have missed?

**Comment by SorraTheOrc** [comment]
Should we always test for exit code? 

Yes

If so, what should be action for a passed and failed validation? For example: 
If passed, should we print anything, or just move on

Just Move on

If failed, should we exit? Should we print an error and move on?

Depends on the setting of --fastfail (set to true, then print error and exit, if false print error and continue)

Do we want to allow wildcards/regex? That could get tricky, but powerful

Start with the simple form of expected similarity (as currently implemented). This has served well with the exception of long bodies of output.

Add wildcard to accommodate long body output (e.g. 'Want to match this content * don't care about what's between, but do care about this')

This will cover 95% of the use cases I've encountered. To cover the rest we want regular expressions, but I don't think we need that until that 5% becomes critical.

How do we want to customize the expected_similarity? (i.e. in the language type? In some other format?)

The current model works well (in the language type). It does not pollute the markdown output and document writers don't need to worry about it, only testers.

Anything else I might have missed?

Don't know! But I'd argue that the existing approach is not broken but can be enhanced through a) refactoring the code and b) allowing wildcards.
